classes/local/attempt_ui/qpy_file_upload.php

```diff
@@ -0,0 +1,170 @@
+        $coursemaxbytes = 0;
+        if (!empty($PAGE->course->maxbytes)) {
+            $coursemaxbytes = $PAGE->course->maxbytes;
+        }
+        $maxbytes = get_user_max_upload_file_size($context, $CFG->maxbytes, $coursemaxbytes, $maxbytes);
+
+        $areamaxbytes = $element->getAttribute('max_bytes_total');
+        $areamaxbytes = is_numeric($areamaxbytes) ? intval($areamaxbytes) : FILE_AREA_MAX_BYTES_UNLIMITED;
+
+        return new static(
+            $element->ownerDocument,
+            $name,
+            $maxfiles,
+            $maxbytes,
+            $areamaxbytes
+        );
+    }
+
+    /**
+     * Renders this element to a DOMNode.
+     *
+     * @param question_attempt $qa
+     * @param question_ui_renderer $renderer
+     * @return DOMNode
+     * @throws coding_exception
+     * @throws file_exception
+     * @throws moodle_exception
+     * @throws stored_file_creation_exception
+     */
+    public function render(question_attempt $qa, question_ui_renderer $renderer): DOMNode {
+        if ($renderer->options->readonly) {
+            return $this->render_readonly($qa, $renderer);
+        } else {
+            return $this->render_writable($qa, $renderer);
+        }
+    }
+
+    /**
+     * Renders a Moodle file manager from this `<qpy:file-upload/>`, preparing it with the last submitted files.
+     *
+     * @param question_attempt $qa
+     * @param question_ui_renderer $renderer
+     * @return DOMNode
+     * @throws coding_exception
+     * @throws file_exception
+     * @throws moodle_exception
+     * @throws stored_file_creation_exception
+     */
+    private function render_writable(question_attempt $qa, question_ui_renderer $renderer): DOMNode {
+        // Re: "global $PAGE cannot be used in renderers" - We're not _that_ kind of a renderer.
+        // phpcs:disable moodle.PHP.ForbiddenGlobalUse.BadGlobal
+        global $CFG, $PAGE, $USER;
+        require_once($CFG->libdir . '/form/filemanager.php');
+
+        $draftitemid = file_get_unused_draft_itemid();
+        $afs = di::get(attempt_file_service::class);
+        $afs->prepare_draft_area($renderer->options->context->id, $qa, $this->name, $USER->id, $draftitemid);
+
+        // TODO: Explain.
+        $renderer->draftareas[$this->name] = $draftitemid;
+
+        $fm = new form_filemanager((object)[
+            'itemid' => $draftitemid,
+            'subdirs' => false,
+            'context' => $renderer->options->context,
+            'maxfiles' => $this->maxfiles,
+            'maxbytes' => $this->maxbytes,
+            'areamaxbytes' => $this->areamaxbytes,
+        ]);
+
+        // phpcs:disable moodle.PHP.ForbiddenGlobalUse.BadGlobal
+        $filesrenderer = $PAGE->get_renderer('core', 'files');
+        $html = $filesrenderer->render($fm);
+        return dom_utils::html_to_fragment($this->doc, $html);
+    }
+
+    private function render_readonly(question_attempt $qa, question_ui_renderer $renderer): DOMNode {
```

Comment:
Hier wäre es besser, einen Moodle-Renderer zu benutzen, damit Themes die Möglichkeit zu Anpassungen haben. Dann können wir diese Ansicht auch in assignsubmission_qpy praktischerweise verwenden.